### PR TITLE
Directional derivative

### PR DESCRIPTION
**Pull request by unnonouno**
I fixed the current check_backward method with so called directional derivative.
This PR include #3216, #3542 and #3550

### CODE REVIEW

**Comment by niboshi** [comment]
Please fix test failures.

**Comment by niboshi** [comment]
Sorry, but can you fix conflict again?

**Comment by niboshi** [comment]
Please fix test errors.

**Comment by niboshi** [comment]
At least the following tests have failed:
* TestScatterAdd_param_4.test_backward_cpu
* TestLinear_param_5.test_backward_cpu
* TestCheckBackward_param_1.test_multiple_output

**Comment by niboshi** [comment]
Please fix `TestLinear_param_4.test_backward_cpu`.

File: chainer/gradient_check.py [hunk 1]
```diff
@@ -138,25 +143,29 @@ def check_backward(func, x_data, y_grad, params=(),
...
+    by comparing this equation with the gradient of :math:`g` numerically
+    calculated and that of :math:`f` computed by backprop.
+    If :math:`r` is chosen from uniform distribution, we can conclude with
+    high probability that the gradient of :math:`f` itself is correct.
 
     If input objects (``x1_data`` or/and ``x2_data`` in this example) represent
     integer variables, their gradients are ignored.
 
-    You can simplify a test when ``MyFunc`` gets only one argument::
+   You can simplify a test when ``MyFunc`` gets only one argument::
```
Comment (hvy) [inline chainer/gradient_check.py]: One less leading whitespace than expected, causing docs to fail.

**Comment by hvy** [comment]
Noticed a small error in the docs. Please see comment.

**Comment by unnonouno** [comment]
fixed

File: chainer/gradient_check.py [hunk 2]
```diff
@@ -138,20 +143,24 @@ def check_backward(func, x_data, y_grad, params=(),
...
-    :math:`g(\\alpha) = f(\\alpha x)`, where :math:`\\alpha \\in \\mathbb{R}`
-    and :math:`f` is a function which actually
-    you want to test.
+    To reduce computational time, it uses directional derivative along a
+    random vector. A function
+    :math:`g: \\mathbb{R} \\rightarrow \\mathbb{R}^n` is defined as
+    :math:`g(\\delta) = f(x + \\delta r)`, where
+    :math:`\\delta \\in \\mathbb{R}`, :math:`r \\in \\mathbb{R}^n`
+    is a random vector
+    and :math:`f` is a function which actually you want to test.
```
Comment (hvy) [inline chainer/gradient_check.py]: `... a function which you actually want ...` (or remove `actually`, completely, I think this latter is cleaner).

**Comment by hvy** [comment]
I am not sure about the correctness of the distribution of sampled offsets but I verified the norms of them and ran tests locally, which seemed fine. 

Other than that, added a few comments regarding docs. I think we also should elaborate on the implications of `no_grad`, e.g. that inputs that are flagged with `no_grads` are basically treated as scalars (not taken into account during the offset normalization). At the moment, this is not obvious by reading the docs.

**Comment by unnonouno** [comment]
Sorry i cann't understand your comment. Input arguments flagged with `no_grads` are not used. So their gradients themselves are not tested.

**Comment by hvy** [comment]
@unnonouno What I basically meant was that the documentation of the `no_grad` argument is a bit poor. But maybe that was because of my lack of understanding.

**Comment by takagi** [comment]
jenkins, test this please.

File: chainer/gradient_check.py [hunk 3]
```diff
@@ -287,76 +278,74 @@ def check_backward(func, x_data, y_grad, params=(),
...
+        # When there is no float variables, we need not to check gradient
+        # values
+        return
+
+    variables = _filter_list(xs, no_grads) + list(params)
     # Keep the gradient arrays of params which may be overwritten by func
-    params_grad = [param.grad for param in params]
-    param_data = [p.data for p in params]
-    casted_data = [x.data.copy() for x in casted_xs]
+    grads = [x.grad for x in variables]
```
Comment (takagi) [inline chainer/gradient_check.py]: Is keeping the references of `x.grad` enough if they are overwritten? I suppose that overwritten are the values of `x.grad`.

File: chainer/gradient_check.py [hunk 4]
```diff
@@ -287,76 +278,74 @@ def check_backward(func, x_data, y_grad, params=(),
...
+
+    if dtype is None:
+        casted_data = [x.data for x in variables]
+    else:
+        if numpy.dtype(dtype).kind != 'f':
```
Comment (takagi) [inline chainer/gradient_check.py]: This check is also on the beginning of the method. We can remove it here.

**Comment by takagi** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
Would you merge this to your repo? https://github.com/unnonouno/chainer/pull/4

**Comment by niboshi** [comment]
LGTM!

**Pull request merged**